src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h

```diff
@@ -130,6 +150,266 @@ namespace OpenMS
+      UInt n_points = 0;
+      for (auto it=p.PosBegin(left); it!=p.PosEnd(right); ++it, ++n_points)
+      {
+        if (peak_height_ < it->getIntensity())
+        {
+          peak_height_ = it->getIntensity();
+          peak_apex_rt_ = it->getPos();
+        }
+      }
+
+      if (integration_type_ == "trapezoid")
+      {
+        for (auto it=p.PosBegin(left); it!=p.PosEnd(right)-1; ++it)
+        {
+          peak_area_ += ((it+1)->getPos() - it->getPos()) * ((it->getIntensity() + (it+1)->getIntensity()) / 2.0);
+        }
+      }
+      else if (integration_type_ == "simpson")
+      {
+        if (n_points < 3)
+        {
+          LOG_DEBUG << std::endl << "Error in integratePeak: number of points must be >=3 for Simpson's rule" << std::endl;
+          return;
+        }
+        if (n_points % 2)
+        {
+          peak_area_ = simpson(p.PosBegin(left), p.PosEnd(right));
+        }
+        else
+        {
+          double areas[4] = {};
+          areas[0] = simpson(p.PosBegin(left), p.PosEnd(right) - 1);   // without last point
+          areas[1] = simpson(p.PosBegin(left) + 1, p.PosEnd(right));   // without first point
+          if (p.begin() <= p.PosBegin(left) - 1)
+          {
+            areas[2] = simpson(p.PosBegin(left) - 1, p.PosEnd(right)); // with one more point on the left
+          }
+          if (p.PosEnd(right) < p.end())
+          {
+            areas[3] = simpson(p.PosBegin(left), p.PosEnd(right) + 1); // with one more point on the right
+          }
+          UInt valids = 0;
+          for (auto area : areas)
+          {
+            if (area)
+            {
+              peak_area_ += area;
+              ++valids;
+            }
+          }
+          peak_area_ /= valids;
+        }
+      }
+      else
+      {
+        std::cout << std::endl << "WARNING: intensity_sum method is being used." << std::endl;
+        for (auto it=p.PosBegin(left); it!=p.PosEnd(right); ++it)
+        {
+          peak_area_ += it->getIntensity();
+        }
+      }
+    }
+
+    template<class PeakContainerConstIteratorT>
+    double simpson_(PeakContainerConstIteratorT it_begin, PeakContainerConstIteratorT it_end) const
+    {
+      double integral = 0.0;
+      for (auto it=it_begin+1; it<it_end-1; it=it+2)
+      {
+        const double h = it->getPos() - (it-1)->getPos();
+        const double k = (it+1)->getPos() - it->getPos();
+        const double y_h = (it-1)->getIntensity();
+        const double y_0 = it->getIntensity();
+        const double y_k = (it+1)->getIntensity();
+        integral += (1.0/6.0) * (h+k) * ((2.0-k/h)*y_h + (pow(h+k,2)/(h*k))*y_0 + (2.0-h/k)*y_k);
+      }
+      return integral;
+    }
+
+    template<class PeakContainerT>
+    void calculatePeakShapeMetrics_(
+      const PeakContainerT& p,
+      const double& left,
+      const double& right,
+      PeakShapeMetrics_& peakShapeMetrics
+    )
+    {
+      peakShapeMetrics.points_across_baseline = 0;
```

Comment:
Do you think it is worth making these class attributes and doing away with the struct?

Reply:
Talking with Timo, he showed me a way to wrap a struct, here: https://github.com/OpenMS/OpenMS/blob/ec89b2f6406d19ba41d0137e64f6e7396736fffc/src/pyOpenMS/pxds/ConsensusFeature.pxd#L83
And the struct is here (looks like a class):
https://github.com/OpenMS/OpenMS/blob/ec89b2f6406d19ba41d0137e64f6e7396736fffc/src/openms/include/OpenMS/KERNEL/ConsensusFeature.h#L116
I'd first try this way and then see how things work. What do you think?

Reply:
It could work